File: java/src/main/java/com/bilt/pos/nexo/security/SaleToPOISecuredMessage.java
/*
 *    ____  _ _ _
 *   | __ )(_) | |_
 *   |  _ \| | | __|
 *   | |_) | | | |_
 *   |____/|_|_|\__|
 *
 *   Bilt POS SDK
 */
package com.bilt.pos.nexo.security;

import com.bilt.pos.nexo.model.ContentInformationType;
import com.bilt.pos.nexo.model.EnvelopedData;
import com.bilt.pos.nexo.model.MessageHeader;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * An encrypted Nexo Sale to POI message using CMS structures.
 *
 * <p>The original JSON payload is AES-encrypted into the {@link EnvelopedData}
 * (with a per-message session key wrapped by the KEK), the unencrypted
 * {@link MessageHeader} is preserved for routing, and a {@link ContentInformationType}
 * security trailer carries the HMAC for integrity verification.</p>
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class SaleToPOISecuredMessage {

    private MessageHeader messageHeader;
    private EnvelopedData envelopedData;
    private ContentInformationType securityTrailer;

    @JsonProperty("MessageHeader")
    public MessageHeader getMessageHeader() { return messageHeader; }
    @JsonProperty("MessageHeader")
    public void setMessageHeader(MessageHeader value) { this.messageHeader = value; }

    @JsonProperty("EnvelopedData")
    public EnvelopedData getEnvelopedData() { return envelopedData; }
    @JsonProperty("EnvelopedData")
    public void setEnvelopedData(EnvelopedData value) { this.envelopedData = value; }

    @JsonProperty("SecurityTrailer")
    public ContentInformationType getSecurityTrailer() { return securityTrailer; }
    @JsonProperty("SecurityTrailer")
    public void setSecurityTrailer(ContentInformationType value) { this.securityTrailer = value; }
}

File: kotlin/src/main/kotlin/com/bilt/pos/nexo/security/SaleToPOISecuredMessage.kt
/*
 *    ____  _ _ _
 *   | __ )(_) | |_
 *   |  _ \| | | __|
 *   | |_) | | | |_
 *   |____/|_|_|\__|
 *
 *   Bilt POS SDK
 */
package com.bilt.pos.nexo.security

import com.bilt.pos.nexo.model.ContentInformationType
import com.bilt.pos.nexo.model.EnvelopedData
import com.bilt.pos.nexo.model.MessageHeader
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * An encrypted Nexo Sale to POI message using CMS structures.
 *
 * The original JSON payload is AES-encrypted into the [EnvelopedData]
 * (with a per-message session key wrapped by the KEK), the unencrypted
 * [MessageHeader] is preserved for routing, and a [ContentInformationType]
 * security trailer carries the HMAC for integrity verification.
 */
@Serializable
data class SaleToPOISecuredMessage(
    @SerialName("MessageHeader")
    val messageHeader: MessageHeader,
    @SerialName("EnvelopedData")
    val envelopedData: EnvelopedData? = null,
    @SerialName("SecurityTrailer")
    val securityTrailer: ContentInformationType? = null
)
